tests/Unit/FileStorageTest.php

```diff
@@ -131,4 +132,26 @@ public function testWriteMetadata(string $metadataClass, string $role, string $e
         $this->expectExceptionMessage('Could not load root metadata.');
         $storage->getRoot();
     }
+
+    public function testFileNamesAreUrlEncoded(): void
+    {
+        $roleName = 'hello/../there!';
+        $fileContents = '{["A test of your mettle"]}';
+
+        $dir = sys_get_temp_dir();
+        $storage = new FileStorage($dir);
+
+        $metadata = $this->prophesize(MetadataBase::class);
+        $metadata->getRole()->willReturn($roleName);
+        $metadata->getSource()->willReturn($fileContents);
+        $metadata->ensureIsTrusted()->shouldBeCalled();
+        $storage->save($metadata->reveal());
+
+        $expectedFileName = urlencode($roleName) . '.json';
```

Comment:
IMHO this is much clearer:
```suggestion
        $expectedFileName = 'hello%2F..%2Fthere.json';
```

(Whenever possible, make test expectations have no logic, so that the before vs after is clearly visible.)